src/node_crypto.cc

```diff
@@ -1025,6 +1066,46 @@ void SecureContext::LoadPKCS12(const FunctionCallbackInfo<Value>& args) {
 }
 
 
+#ifndef OPENSSL_NO_ENGINE
+void SecureContext::SetClientCertEngine(
+  const FunctionCallbackInfo<Value>& args) {
+  Environment* env = Environment::GetCurrent(args);
+  CHECK(args.Length() == 1 && args[0]->IsString());
+
+  SecureContext* sc = Unwrap<SecureContext>(args.This());
+
+  ClearErrorOnReturn clear_error_on_return;
+
+  // If an engine was previously set, free it.
+  // This is because SSL_CTX_set_client_cert_engine does not itself
+  // 'support' multiple calls by cleaning up before overwriting the
+  // client_cert_engine internal context variable
+  if (sc->ctx_->client_cert_engine != NULL) {
+    ENGINE* e = sc->ctx_->client_cert_engine;
+    sc->ctx_->client_cert_engine = nullptr;
+    ENGINE_finish(e);
+  }
+
+  // Load engine
+  const node::Utf8Value engine_id(env->isolate(), args[0]);
+  ENGINE* engine = LoadEngineById(env, *engine_id);
+
+  if (engine == nullptr) {
+    // Load failed... appropriate js exception has been scheduled
+    return;
+  }
+
+  int r = SSL_CTX_set_client_cert_engine(sc->ctx_, engine);
+  // Free reference (SSL_CTX_set_client_cert_engine took it via ENGINE_init)
+  ENGINE_free(engine);
+  if (r == 0) {
+    ERR_clear_error();
+    return ThrowCryptoError(env, ERR_get_error());
```

Comment:
`ERR_get_error()` after `ERR_clear_error()` can't possibly do the right thing... can you drop the `ERR_clear_error()` and replace the `ClearErrorOnReturn` above with a `MarkPopErrorOnReturn`?